File: Batch2/Trees/Binary Tree/Introduction/Traversal.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Traversal {
    class Pair {
        TreeNode node;
        int state;

        public Pair(TreeNode node, int state){
            this.node = node;
            this.state = state;
        }
    }

    public static void traverse(TreeNode root){
        Stack<Pair> st = new Stack<>();

        if(root != null){
            st.push(new Pair(root,0));
        }

        while(st.size() > 0){
            TreeNode topNode = st.peek().node;

            if(st.peek().state == 0){ // seeing node for the first time, push left
                preorder.add(topNode.val);
                st.peek().state++;

                if(topNode.left != null){
                    st.push(new Pair(topNode.left,0));
                }

            } else if(st.peek().state == 1){ // already seen left subtree, push right
                inorder.add(topNode.val);
                st.peek().state++;

                if(topNode.right != null){
                    st.push(new Pair(topNode.right, 0));
                }
            } else { // seen both left and right subtree, pop
                postorder.add(topNode.val);

                st.pop();
            }
        }
    }

    public static void main(String[] args){

    }
}

File: Batch2/Trees/Binary Tree/Introduction/traversal.cpp
#include<bits/stdc++.h>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void traversal(TreeNode* root){
    stack<pair<TreeNode*,int>> st;
    vector<int> preorder;
    vector<int> inorder;
    vector<int> postorder;

    if(root){
        st.push({root,0});
    }

    while(st.size() > 0){
        TreeNode topNode = st.top().first;

        if(st.top().second == 0){
            preorder.push_back(topNode->val);
            
            st.top().second++;

            if(topNode->left){
                st.push({topNode->left, 0});
            }
        } else if(st.top().second == 1){
            inorder.push_back(topNode->val);

            st.top().second++;

            if(topNode->right){
                st.push({topNode->right,0});
            } 
        } else {
            postorder.push_back(topNode->val);

            st.pop();
        }
    }
}